File: java/src/main/java/shared/javaUtil/Exposure.java
package shared.javaUtil;

import java.util.function.Supplier;
import shared.javaUtil.enums.ExposureCategory;
import shared.javaUtil.enums.ExposureLevel;
import static shared.javaUtil.enums.ExposureLevel.*;

/**
 * An instance-bound facade and DSL handle for {@link Expresso} logging routines.
 * <p>
 * An {@code Exposure} handle encapsulates an optional identity tag (e.g., component or class name)
 * and provides shorthand methods ({@code l1}–{@code l5}, {@code err}, {@code here}) for logging events
 * with automatic identity tag injection and hot-loop safe lazy message evaluation.
 */
public final class Exposure {

    private final String identity;

    /**
     * Constructs a default package-private {@code Exposure} instance without an assigned identity tag.
     */
    Exposure() { this(null); }

    /**
     * Constructs a package-private {@code Exposure} instance bound to a specific component identity.
     *
     * @param identity The identifier tag representing the source component (e.g., "RenderEngine").
     */
    Exposure(String identity) { this.identity = identity; }

    /**
     * Conditionally formats and emits a log message using the bound identity context.
     *
     * @param category The logging category.
     * @param level    The required exposure level.
     * @param format   The string format template.
     * @param args     The format template arguments.
     */
    public void expose(ExposureCategory category, ExposureLevel level, String format, Object... args) {
        Expresso.printf(identity, category, level, format, args);
    }
/**
     * Shorthand for logging a formatted message at {@link ExposureLevel#LEVEL1}.
     *
     * @param category The logging category.
     * @param format   The string format template.
     * @param args     The format template arguments.
     */
    public void l1(ExposureCategory category, String format, Object... args) {
        Expresso.printf(identity, category, LEVEL1, format, args);
    }

    /**
     * Shorthand for logging a formatted message at {@link ExposureLevel#LEVEL2}.
     *
     * @param category The logging category.
     * @param format   The string format template.
     * @param args     The format template arguments.
     */
    public void l2(ExposureCategory category, String format, Object... args) {
        Expresso.printf(identity, category, LEVEL2, format, args);
    }

    /**
     * Shorthand for logging a formatted message at {@link ExposureLevel#LEVEL3}.
     *
     * @param category The logging category.
     * @param format   The string format template.
     * @param args     The format template arguments.
     */
    public void l3(ExposureCategory category, String format, Object... args) {
        Expresso.printf(identity, category, LEVEL3, format, args);
    }

    /**
     * Shorthand for logging a formatted message at {@link ExposureLevel#LEVEL4}.
     *
     * @param category The logging category.
     * @param format   The string format template.
     * @param args     The format template arguments.
     */
    public void l4(ExposureCategory category, String format, Object... args) {
        Expresso.printf(identity, category, LEVEL4, format, args);
    }

    /**
     * Shorthand for logging a formatted message at {@link ExposureLevel#LEVEL5}.
     *
     * @param category The logging category.
     * @param format   The string format template.
     * @param args     The format template arguments.
     */
    public void l5(ExposureCategory category, String format, Object... args) {
        Expresso.printf(identity, category, LEVEL5, format, args);
    }

    // --- Lazy Supplier Evaluation Logging Methods ---

    /**
     * Conditionally evaluates and emits a lazy log message using the bound identity context.
     * <p>
     * The {@link Supplier} is evaluated only if the category and level pass active exposure filters,
     * preventing unnecessary string allocations in performance-critical code paths.
     *
     * @param category The logging category.
     * @param level    The required exposure level.
     * @param message  A supplier yielding the log string.
     */
    public void expose(ExposureCategory category, ExposureLevel level, Supplier<String> message) {
        if (!Expresso.isExposed(category, level)) return;
        Expresso.printf(identity, category, level, "%s", message.get());
    }

    /**
     * Lazily evaluates and logs a message at {@link ExposureLevel#LEVEL1}.
     *
     * @param category The logging category.
     * @param message  A supplier yielding the log string.
     */
    public void l1(ExposureCategory category, Supplier<String> message) {
        if (!Expresso.isExposed(category, LEVEL1)) return;
        Expresso.printf(identity, category, LEVEL1, "%s", message.get());
    }

    /**
     * Lazily evaluates and logs a message at {@link ExposureLevel#LEVEL2}.
     *
     * @param category The logging category.
     * @param message  A supplier yielding the log string.
     */
    public void l2(ExposureCategory category, Supplier<String> message) {
        if (!Expresso.isExposed(category, LEVEL2)) return;
        Expresso.printf(identity, category, LEVEL2, "%s", message.get());
    }

    /**
     * Lazily evaluates and logs a message at {@link ExposureLevel#LEVEL3}.
     *
     * @param category The logging category.
     * @param message  A supplier yielding the log string.
     */
    public void l3(ExposureCategory category, Supplier<String> message) {
        if (!Expresso.isExposed(category, LEVEL3)) return;
        Expresso.printf(identity, category, LEVEL3, "%s", message.get());
    }

    /**
     * Lazily evaluates and logs a message at {@link ExposureLevel#LEVEL4}.
     *
     * @param category The logging category.
     * @param message  A supplier yielding the log string.
     */
    public void l4(ExposureCategory category, Supplier<String> message) {
        if (!Expresso.isExposed(category, LEVEL4)) return;
        Expresso.printf(identity, category, LEVEL4, "%s", message.get());
    }

    /**
     * Lazily evaluates and logs a message at {@link ExposureLevel#LEVEL5}.
     *
     * @param category The logging category.
     * @param message  A supplier yielding the log string.
     */
    public void l5(ExposureCategory category, Supplier<String> message) {
        if (!Expresso.isExposed(category, LEVEL5)) return;
        Expresso.printf(identity, category, LEVEL5, "%s", message.get());
    }

    // --- Helpers & Execution Utilities ---

    /**
     * Checks whether a specific category and level combination is currently active in Expresso.
     *
     * @param category The category to evaluate.
     * @param level    The level to evaluate.
     * @return {@code true} if exposed, {@code false} otherwise.
     */
    public boolean enabled(ExposureCategory category, ExposureLevel level) { 
        return Expresso.isExposed(category, level); 
    }

    /**
     * Emits a double Expresso.NEWLINE to output if the category and level criteria are met.
     *
     * @param category      The target category.
     * @param requiredLevel The required severity level.
     */
    public void nl(ExposureCategory category, ExposureLevel requiredLevel) {
        Expresso.NEWLINE(category, requiredLevel);
    }

    /**
     * Emits a visual Expresso.LINEBREAK separator if the category and level criteria are met.
     *
     * @param category      The target category.
     * @param requiredLevel The required severity level.
     */
    public void lbr(ExposureCategory category, ExposureLevel requiredLevel) {
        Expresso.LINEBREAK(category, requiredLevel);
    }

    /**
     * Prints an unrestricted Expresso.error message. Prefixes the format template with 
     * the bound identity tag if present.
     *
     * @param format The format template.
     * @param args   Arguments or optional trailing {@link Throwable}.
     */
    public void err(String format, Object... args) { 
        if (identity != null && !identity.isBlank()) {
            Expresso.error("[" + identity + "] " + format, args);
        } else {
            Expresso.error(format, args);
        }
    }

    /**
     * Lazily evaluates and prints an unrestricted Expresso.error message with the bound identity tag if present.
     *
     * @param message A supplier yielding the Expresso.error string.
     */
    public void err(Supplier<String> message) {
        if (message == null) return;
        if (identity != null && !identity.isBlank()) {
            Expresso.error("[" + identity + "] " + message.get());
        } else {
            Expresso.error(message.get());
        }
    }

    /**
     * Rapid debugging utility that prints a high-visibility execution marker tag, 
     * automatically injecting the bound identity tag if present.
     */
    public void here() { 
        if (identity != null && !identity.isBlank()) {
            Expresso.hereAnnounce(identity);
        } else {
            Expresso.hereAnnounce(); 
        }
    }

    /**
     * Rapid debugging utility that prints a high-visibility execution marker tag 
     * along with a custom label, automatically injecting the bound identity tag if present.
     *
     * @param label Additional context or descriptive text.
     */
    public void here(String label) { 
        if (identity != null && !identity.isBlank()) {
            Expresso.hereAnnounce(identity, label);
        } else {
            Expresso.hereAnnounce("", label);
        }
    }

    /**
     * Rapid debugging utility allowing a single-use custom color and header override.
     * 
     * @param customColor  single-use ANSI color code.
     * @param customHeader single-use marker header string.
     * @param label        Additional context or identifying text.
     */
    public void here(String customColor, String customHeader, String label) { 
        if (identity != null && !identity.isBlank()) {
            Expresso.hereAnnounce(identity, customColor, customHeader, label);
        } else {
            Expresso.hereAnnounce(customColor, customHeader, label);
        }
    }

    /**
     * Retrieves the identity tag bound to this Exposure instance.
     *
     * @return The identity string, or {@code null} if no identity was set.
     */
    public String getIdentity() { return identity; }
}

File: c/src/Exposure.c
#include <stdio.h>
#include <stdarg.h>

#include "expresso/Exposure.h"
#include "expresso/Expresso.h"

#define EXPOSURE_BUFFER_SIZE 4096

Exposure create(void) {
    Exposure exposure = {NULL};
    return exposure;
}

Exposure createWithIdentity(const char *identity) {
    Exposure exposure = {identity};
    return exposure;
}

/**
 * Conditionally formats and emits an exposure message using the bound identity.
 *
 * The message is emitted only when the specified category and exposure level
 * are currently enabled.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param level The required exposure level.
 * @param format The message format string.
 * @param ... Arguments used by the format string.
 */
void expose(Exposure *exposure, ExposureCategory category, ExposureLevel level, const char *format, ...) {
    if (exposure == NULL || format == NULL) return;
    if (!isExposed(category, level)) return;

    char message[EXPOSURE_BUFFER_SIZE];

    va_list arguments;
    va_start(arguments, format);
    vsnprintf(message, sizeof(message), format, arguments);
    va_end(arguments);

    expressoPrintf(exposure->identity, category, level, "%s", message);
}

/**
 * Shorthand for exposing a formatted message at LEVEL1.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param format The message format string.
 * @param ... Arguments used by the format string.
 */
void l1(Exposure *exposure, ExposureCategory category, const char *format, ...) {
    if (exposure == NULL || format == NULL) return;
    if (!isExposed(category, LEVEL1)) return;

    char message[EXPOSURE_BUFFER_SIZE];

    va_list arguments;
    va_start(arguments, format);
    vsnprintf(message, sizeof(message), format, arguments);
    va_end(arguments);

    expressoPrintf(exposure->identity, category, LEVEL1, "%s", message);
}

/**
 * Shorthand for exposing a formatted message at LEVEL2.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param format The message format string.
 * @param ... Arguments used by the format string.
 */
void l2(Exposure *exposure, ExposureCategory category, const char *format, ...) {
    if (exposure == NULL || format == NULL) return;
    if (!isExposed(category, LEVEL2)) return;

    char message[EXPOSURE_BUFFER_SIZE];

    va_list arguments;
    va_start(arguments, format);
    vsnprintf(message, sizeof(message), format, arguments);
    va_end(arguments);

    expressoPrintf(exposure->identity, category, LEVEL2, "%s", message);
}

/**
 * Shorthand for exposing a formatted message at LEVEL3.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param format The message format string.
 * @param ... Arguments used by the format string.
 */
void l3(Exposure *exposure, ExposureCategory category, const char *format, ...) {
    if (exposure == NULL || format == NULL) return;
    if (!isExposed(category, LEVEL3)) return;

    char message[EXPOSURE_BUFFER_SIZE];

    va_list arguments;
    va_start(arguments, format);
    vsnprintf(message, sizeof(message), format, arguments);
    va_end(arguments);

    expressoPrintf(exposure->identity, category, LEVEL3, "%s", message);
}

/**
 * Shorthand for exposing a formatted message at LEVEL4.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param format The message format string.
 * @param ... Arguments used by the format string.
 */
void l4(Exposure *exposure, ExposureCategory category, const char *format, ...) {
    if (exposure == NULL || format == NULL) return;
    if (!isExposed(category, LEVEL4)) return;

    char message[EXPOSURE_BUFFER_SIZE];

    va_list arguments;
    va_start(arguments, format);
    vsnprintf(message, sizeof(message), format, arguments);
    va_end(arguments);

    expressoPrintf(exposure->identity, category, LEVEL4, "%s", message);
}

/**
 * Shorthand for exposing a formatted message at LEVEL5.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param format The message format string.
 * @param ... Arguments used by the format string.
 */
void l5(Exposure *exposure, ExposureCategory category, const char *format, ...) {
    if (exposure == NULL || format == NULL) return;
    if (!isExposed(category, LEVEL5)) return;

    char message[EXPOSURE_BUFFER_SIZE];

    va_list arguments;
    va_start(arguments, format);
    vsnprintf(message, sizeof(message), format, arguments);
    va_end(arguments);

    expressoPrintf(exposure->identity, category, LEVEL5, "%s", message);
}

/**
 * @name Lazy Exposure
 * @{
 */

/**
 * Conditionally evaluates and emits a lazy exposure message using the bound identity.
 *
 * The callback is evaluated only when the specified category and exposure level
 * pass the active exposure filters. This avoids unnecessary message construction
 * when the exposure is disabled.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param level The required exposure level.
 * @param message Callback that produces the exposure message.
 */
void exposeLazy(Exposure *exposure, ExposureCategory category, ExposureLevel level, const char *(*message)(void)) {
    if (exposure == NULL || message == NULL) return;
    if (!isExposed(category, level)) return;

    const char *result = message();
    if (result == NULL) return;

    expressoPrintf(exposure->identity, category, level, "%s", result);
}

/**
 * Lazily evaluates and exposes a message at LEVEL1.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param message Callback that produces the exposure message.
 */
void l1Lazy(Exposure *exposure, ExposureCategory category, const char *(*message)(void)) {
    if (exposure == NULL || message == NULL) return;
    if (!isExposed(category, LEVEL1)) return;

    const char *result = message();
    if (result == NULL) return;

    expressoPrintf(exposure->identity, category, LEVEL1, "%s", result);
}

/**
 * Lazily evaluates and exposes a message at LEVEL2.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param message Callback that produces the exposure message.
 */
void l2Lazy(Exposure *exposure, ExposureCategory category, const char *(*message)(void)) {
    if (exposure == NULL || message == NULL) return;
    if (!isExposed(category, LEVEL2)) return;

    const char *result = message();
    if (result == NULL) return;

    expressoPrintf(exposure->identity, category, LEVEL2, "%s", result);
}

/**
 * Lazily evaluates and exposes a message at LEVEL3.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param message Callback that produces the exposure message.
 */
void l3Lazy(Exposure *exposure, ExposureCategory category, const char *(*message)(void)) {
    if (exposure == NULL || message == NULL) return;
    if (!isExposed(category, LEVEL3)) return;

    const char *result = message();
    if (result == NULL) return;

    expressoPrintf(exposure->identity, category, LEVEL3, "%s", result);
}

/**
 * Lazily evaluates and exposes a message at LEVEL4.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param message Callback that produces the exposure message.
 */
void l4Lazy(Exposure *exposure, ExposureCategory category, const char *(*message)(void)) {
    if (exposure == NULL || message == NULL) return;
    if (!isExposed(category, LEVEL4)) return;

    const char *result = message();
    if (result == NULL) return;

    expressoPrintf(exposure->identity, category, LEVEL4, "%s", result);
}

/**
 * Lazily evaluates and exposes a message at LEVEL5.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param message Callback that produces the exposure message.
 */
void l5Lazy(Exposure *exposure, ExposureCategory category, const char *(*message)(void)) {
    if (exposure == NULL || message == NULL) return;
    if (!isExposed(category, LEVEL5)) return;

    const char *result = message();
    if (result == NULL) return;

    expressoPrintf(exposure->identity, category, LEVEL5, "%s", result);
}

/** @} */

/**
 * Checks whether a specific category and exposure level are currently enabled.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param level The exposure level to check.
 * @return true if the category and level are currently exposed; false otherwise.
 */
bool enabled(Exposure *exposure, ExposureCategory category, ExposureLevel level) {
    if (exposure == NULL) return false;
    return isExposed(category, level);
}

/**
 * Emits a blank line when the specified category and exposure level are enabled.
 *
 * This produces the same output as the Expresso newline helper and does not
 * require a message.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param requiredLevel The required exposure level.
 */
void nl(Exposure *exposure, ExposureCategory category, ExposureLevel requiredLevel) {
    if (exposure == NULL) return;
    expressoNewline(category, requiredLevel);
}

/**
 * Emits a visual line-break separator when the specified category and exposure
 * level are enabled.
 *
 * @param exposure The exposure context.
 * @param category The exposure category.
 * @param requiredLevel The required exposure level.
 */
void lbr(Exposure *exposure, ExposureCategory category, ExposureLevel requiredLevel) {
    if (exposure == NULL) return;
    expressoLinebreak(category, requiredLevel);
}

/**
 * Reports an unrestricted Expresso error message.
 *
 * The bound identity is included in the error message when an identity is set.
 * Error messages are not affected by category or exposure-level filters.
 *
 * @param exposure The exposure context.
 * @param format The error message format string.
 * @param ... Arguments used by the format string.
 */
void err(Exposure *exposure, const char *format, ...) {
    if (exposure == NULL || format == NULL) return;

    char message[EXPOSURE_BUFFER_SIZE];

    va_list arguments;
    va_start(arguments, format);
    vsnprintf(message, sizeof(message), format, arguments);
    va_end(arguments);

    if (exposure->identity != NULL && exposure->identity[0] != '\0') {
        error("[%s] %s", exposure->identity, message);
    } else {
        error("%s", message);
    }
}

/**
 * Lazily evaluates and reports an unrestricted Expresso error message.
 *
 * The callback is evaluated immediately because errors are not subject to
 * category or exposure-level filtering.
 *
 * The bound identity is included in the error message when an identity is set.
 *
 * @param exposure The exposure context.
 * @param message Callback that produces the error message.
 */
void errLazy(Exposure *exposure, const char *(*message)(void)) {
    if (exposure == NULL || message == NULL) return;

    const char *result = message();
    if (result == NULL) return;

    if (exposure->identity != NULL && exposure->identity[0] != '\0') {
        error("[%s] %s", exposure->identity, result);
    } else {
        error("%s", result);
    }
}

/**
 * Emits a high-visibility Expresso execution marker.
 *
 * The bound identity is included when an identity is set.
 */
void here(Exposure *exposure) {
    if (exposure == NULL) return;
    hereAnnounce(exposure->identity, NULL);
}

/**
 * Emits a high-visibility Expresso execution marker with a custom label.
 *
 * The bound identity is included when an identity is set.
 *
 * @param exposure The exposure context.
 * @param label Additional context or descriptive text.
 */
void hereLabel(Exposure *exposure, const char *label) {
    if (exposure == NULL) return;
    hereAnnounce(exposure->identity, label);
}

/**
 * Emits a high-visibility Expresso execution marker with custom formatting.
 *
 * The custom color and header apply only to this invocation. The bound identity
 * is included when an identity is set.
 *
 * @param exposure The exposure context.
 * @param customColor ANSI color/style code for the marker.
 * @param customHeader Header text for the marker.
 * @param label Additional context or descriptive text.
 */
void hereCustom(Exposure *exposure, const char *customColor, const char *customHeader, const char *label) {
    if (exposure == NULL) return;
    hereAnnounceCustom(exposure->identity, customColor, customHeader, label);
}

/**
 * Retrieves the identity bound to this Exposure instance.
 *
 * @param exposure The exposure context.
 * @return The identity string, or NULL if no identity is set.
 */
const char *getIdentity(Exposure *exposure) {
    if (exposure == NULL) return NULL;
    return exposure->identity;
}
